### PR TITLE
[SCO-1678] Add programable retry mechanism

### PR DESCRIPTION
**Pull request by beslic**
I have not added any tests as this does not change current logic or behaviour.

### CODE REVIEW

File: pester.go [hunk 1]
```diff
@@ -135,8 +137,12 @@ type ContextLogHook func(ctx context.Context, e ErrEntry)
 // BackoffStrategy is used to determine how long a retry request should wait until attempted
 type BackoffStrategy func(retry int) time.Duration
 
+// RetryStrategy is used for determining when to retry request based on the response status code.
+// If this function returns true, attempt is considered successful and no more attempts are performed.
```
Comment (vpintaric-axilis) [inline pester.go]: Would it make sense to invert the logic here? i.e.
```suggestion
// If this function returns true, we consider that the request has failed and should be retried.
```
(also doing the appropriate code changes, of course)

Makes the code a bit more readable IMO. Currently you've got a `Retry` function which, when it returns `true`, we consider "no more retries" which is a bit counterintuitive. If you go ahead with this change I would also recommend renaming `Retry` member attribute to `ShouldRetry` to make it even more clearer.

File: pester.go [hunk 2]
```diff
@@ -165,6 +171,14 @@ func LinearJitterBackoff(i int) time.Duration {
 	return jitter(i)
 }
 
+// DefaultRetryStrategy returns true in the following case:
+//  - status code is lower than http.StatusInternalServerError and
+//  - status is not http.StatusTooManyRequests or status is http.StatusTooManyRequests with retryOnHTTP429 flag set false
+func DefaultRetryStrategy(status int, retryOnHTTP429 bool) bool {
+	return status < http.StatusInternalServerError &&
+		(status != http.StatusTooManyRequests || (status == http.StatusTooManyRequests && !retryOnHTTP429))
+}
```
Comment (vpintaric-axilis) [inline pester.go]: What do you think about making `RetryStrategy` simply `func(status int) bool` and then changing this to

```suggestion
// MakeDefaultRetryStrategy creates a function which returns true in the following case:
//  - status code is lower than http.StatusInternalServerError and
//  - status is not http.StatusTooManyRequests or status is http.StatusTooManyRequests with retryOnHTTP429 flag set false
func MakeDefaultRetryStrategy(retryOnHTTP429 bool) RetryStrategy {
	return func(status int) bool {
		status < http.StatusInternalServerError &&
			(status != http.StatusTooManyRequests || (status == http.StatusTooManyRequests && !retryOnHTTP429))
	}
}
```

Note - not sure whether I got the syntax right here.

**Comment by vpintaric-axilis** [comment]
> I have not added any tests as this does not change current logic or behaviour.

I believe it does make sense to add a test case where this new feature is used. E.g. add a test case where the client should retry even in case of 404.

File: pester.go (hunk 2 above)
Comment (idljubicic) [reply pester.go:180]: I found this function a bit har to read easily, so what do you thing about switching to simple if cases?

```
// DefaultRetryStrategy returns true in the following case:
//  - status code is lower than http.StatusInternalServerError and status is not http.StatusTooManyRequests
//  - status code is lower than http.StatusInternalServerError and status is http.StatusTooManyRequests 
// 	with retryOnHTTP429 flag set to false
func DefaultRetryStrategy(status int, retryOnHTTP429 bool) bool {
	if status < http.StatusInternalServerError && status != http.StatusTooManyRequests {
		return true
	}
	if status < http.StatusInternalServerError && status == http.StatusTooManyRequests && !retryOnHTTP429 {
		return true
	}
	return false
}
```

File: pester.go (hunk 1 above)
Comment (idljubicic) [reply pester.go]: Retry function returning true when we consider "no more retries" seems counterintuitive to me too, I agree with Vedran.

**Comment by idljubicic** [comment]
> > I have not added any tests as this does not change current logic or behaviour.
> 
> I believe it does make sense to add a test case where this new feature is used. E.g. add a test case where the client should retry even in case of 404.

**Comment by idljubicic** [comment]
Accidental PR closing, my bad :) Reopened it right away.

File: pester.go (hunk 2 above)
Comment (vpintaric-axilis) [reply pester.go]: > what do you thing about switching to simple if cases?

Sounds good to me.
But IMO it should still be:
```
func MakeDefaultRetryStrategy(retryOnHTTP429 bool) RetryStrategy {
	return func(status int) bool {
		// default implementation
	}
}
```

For backwards compatibility reasons and to make sure that public API is _somewhat sane_.

EDIT: I know there's the "what do you thing" typo but I'm gonna keep it ❤️

Comment (idljubicic) [reply pester.go:180]: // DefaultRetryStrategy returns true in the following case:
//  - status code is lower than http.StatusInternalServerError and
//  - status is not http.StatusTooManyRequests or status is http.StatusTooManyRequests with retryOnHTTP429 flag set false
func DefaultRetryStrategy(status int, retryOnHTTP429 bool) bool {
	return status < http.StatusInternalServerError &&
		(status != http.StatusTooManyRequests || (status == http.StatusTooManyRequests && !retryOnHTTP429))
}
```

Comment (idljubicic) [inline pester.go]: I found this function a bit hard to read easily, so what do you think about switching to simple if cases?

```suggestion
// DefaultRetryStrategy returns true in the following case:
//  - status code is lower than http.StatusInternalServerError and status is not http.StatusTooManyRequests
//  - status code is lower than http.StatusInternalServerError and status is http.StatusTooManyRequests 
//    with retryOnHTTP429 flag set to false
func DefaultRetryStrategy(status int, retryOnHTTP429 bool) bool {
	if status < http.StatusInternalServerError && status != http.StatusTooManyRequests {
		return true
	}
	if status < http.StatusInternalServerError && status == http.StatusTooManyRequests && !retryOnHTTP429 {
		return true
	}
	return false
}
```

Comment (idljubicic) [reply pester.go]: I played with Github features so the review currently looks horrible, my updated comment which you replied to is now below your response. Also this "what do you thing" typo is triggering me :D

File: pester.go (hunk 1 above)
Comment (beslic) [reply pester.go]: Yeah, counterintuitive, will switch it back.

File: pester.go (hunk 2 above)
Comment (beslic) [reply pester.go]: I will try that.

Comment (beslic) [reply pester.go]: I tried all possible ways to make this similar to the what Vedran suggested, but failed. It's impossible to make 2 parameter function 1 parameter, and still keep it simple and easy to use.

I still think that one function with 2 parameters and simple bool return is the simplest solution to this problem.

File: pester.go [hunk 3]
```diff
@@ -165,6 +172,19 @@ func LinearJitterBackoff(i int) time.Duration {
...
 
+// DefaultRetryStrategy returns true (request is retried) when the status code is equal to or higher
+// than http.StatusInternalServerError.
+func DefaultRetryStrategy(status int) bool {
+	return status >= http.StatusInternalServerError
+}
+
+// defaultRetryStrategyWith429 returns true (request is retried) when the status code is equal to or
+// higher than http.StatusTooManyRequests or if status is http.StatusTooManyRequests. This strategy
+// is used when setting Client.SetRetryOnHTTP429.
```
Comment (vpintaric-axilis) [inline pester.go]:
```suggestion
// is used when RetryOnHTTP429 is set to `true`.
```

**Review by vpintaric-axilis — APPROVED** [review]

**Review by idljubicic — APPROVED** [review]

**Pull request merged**